File: C++/Dynamic Programming/house-robber-198.cpp
/*
    LeetCode Question 198
    House Robber
*/

#include <vector>

using namespace std;

int rob(vector<int> &nums) {
  int n = nums.size();
  vector<int> dp(n, 0);

  if (n == 0) {
    return 0;
  }
  if (n == 1) {
    return nums[0];
  }
  dp[0] = nums[0];
  dp[1] = max(nums[0], nums[1]);

  // Dynamnic Programming
  // Either rob the current house and the sum of the house twice before
  //        or take the sum of the previous house
  for (int i = 2; i < n; i++) {
    dp[i] = max(dp[i - 1], nums[i] + dp[i - 2]);
  }

  // The last slot in the DP vector should return the max value
  return dp[n - 1];
}

File: Python/Dynamic_Programming/house-robber-198.py
# Bottom Up DP - [Leetcode 198](https://leetcode.com/problems/house-robber/)

from typing import List

def rob(self, nums: List[int]) -> int:
    # 1 2 3 1
    # 1 2 4 4 
    if len(nums) < 2:
        return nums[0]
    dp = []
    dp.append( nums[0] )
    dp.append(max(nums[0], nums[1]))
    for i in range(2, len(nums)):
        dp.append(max( nums[i] + dp[i-2], dp[i-1]) )
    return dp[-1]
